### PR TITLE
Support input_names in dynamic_axes

### LINKED ISSUE #129276: [ONNX] Full support of dynamic axes
Support torch.onnx.export api when the inputs are renamed, and support torch.onnx.dynamo_export api when we use `True` to make everything dynamic without having to specify axis names.

### PR DESCRIPTION
**Pull request by titaiwangms**
Fix https://github.com/pytorch/pytorch/issues/129276

Tested with
```python
    def test_input_names_are_not_yet_supported_in_dynamic_axes(self):
        import torch_onnx
        torch_onnx.patch_torch(error_report=True, profile=True)
        
        a = torch.onnx.export(
            SampleModelForDynamicShapes(),
            (
                torch.randn(2, 2, 3),
                torch.randn(2, 2, 3),
            ),
            input_names=["input"],
            dynamic_axes={"input": [0, 1]},
            dynamo=True,
        )
        a.save("test.onnx")
```
Screenshot 2024-07-05 111822

### CODE REVIEW

File: src/torch_onnx/_patch.py [hunk 1]
```diff
@@ -48,41 +48,48 @@ def _from_dynamic_axes_to_dynamic_shapes(
...
+        if len(input_names) > len(sig.parameters):
+            raise ValueError(
+                f"Number of input names ({len(input_names)}) should not be greater than the number of model inputs ({len(sig.parameters)})"
+            )
+        input_names_to_model_inputs = {
+            input_name: param_name
+            for input_name, param_name in zip(input_names, sig.parameters)
+        }
+
+    # NOTE: torch.export.export does not support input names assignmen,
```
Comment (justinchuby) [inline src/torch_onnx/_patch.py]:
```suggestion
    # NOTE: torch.export.export does not support input names assignment,
```

File: src/torch_onnx/_patch.py [hunk 2]
```diff
@@ -48,41 +48,48 @@ def _from_dynamic_axes_to_dynamic_shapes(
...
+    # so we need to map input names to model inputs to create dynamic_shapes
+    # for the exported program
+    dynamic_shapesto_exported_program = {}
```
Comment (justinchuby) [inline src/torch_onnx/_patch.py]:
```suggestion
    dynamic_shapes_to_exported_program = {}
```

File: src/torch_onnx/_patch.py [hunk 3]
```diff
@@ -48,41 +48,48 @@ def _from_dynamic_axes_to_dynamic_shapes(
...
     for input_name, axes in dynamic_axes.items():
-        if input_name in input_names_set:
+        if input_name not in input_names_to_model_inputs:
             raise ValueError(
-                "input names is not supported yet. Please use model forward signature."
+                f"dynamix axe: {input_name} is not found in the input names: {input_names}"
```
Comment (justinchuby) [inline src/torch_onnx/_patch.py]:
```suggestion
                dynamic axis: '{input_name}' is not found in the input names: {input_names}"
```

File: src/torch_onnx/_patch.py (hunk 2 above)
Comment (justinchuby) [reply src/torch_onnx/_patch.py]: or: to... -> for_torch_export?

**Review by justinchuby — APPROVED** [review]
Thank you!

File: src/torch_onnx/_patch.py [hunk 4]
```diff
@@ -48,41 +48,48 @@ def _from_dynamic_axes_to_dynamic_shapes(
...
             )
+        model_input_name = input_names_to_model_inputs[input_name]
```
Comment (justinchuby) [inline src/torch_onnx/_patch.py:79]: So model input is the name presented in the forward function right? Just checking my understanding

File: src/torch_onnx/_patch.py [hunk 5]
```diff
@@ -48,41 +48,48 @@ def _from_dynamic_axes_to_dynamic_shapes(
     """
     # https://github.com/pytorch/pytorch/pull/128371
     if dynamic_axes is None:
-        return None
+        return None, input_names
```
Comment (justinchuby) [inline src/torch_onnx/_patch.py]: Is returning input_names expected? Looks like we expect that to be `dict[str, Any] | None`

File: src/torch_onnx/_patch.py (hunk 4 above)
Comment (titaiwangms) [reply src/torch_onnx/_patch.py:79]: Yes

File: src/torch_onnx/_patch.py (hunk 5 above)
Comment (titaiwangms) [reply src/torch_onnx/_patch.py]: Done

File: src/torch_onnx/_patch.py [hunk 6]
```diff
@@ -50,39 +50,46 @@ def _from_dynamic_axes_to_dynamic_shapes(
...
+            for input_name, param_name in zip(input_names, sig.parameters)
+        }
+
+    # NOTE: torch.export.export does not support input names assignment,
+    # so we need to map input names to model inputs to create dynamic_shapes
+    # for the exported program
+    dynamic_shapes_to_exported_program = {}
     for input_name, axes in dynamic_axes.items():
-        if input_name in input_names_set:
+        if input_name not in input_names_to_model_inputs:
```
Comment (justinchuby) [inline src/torch_onnx/_patch.py:75]: When input_names is None, will this always raise the ValueError below? (also input_names_to_model_inputs looks to be uninitialized when input_names is None)

File: src/torch_onnx/_patch.py [hunk 7]
```diff
@@ -47,42 +47,57 @@ def _from_dynamic_axes_to_dynamic_shapes(
 
     """
     # https://github.com/pytorch/pytorch/pull/128371
+
```
Comment (justinchuby) [inline src/torch_onnx/_patch.py]:
```suggestion

```

File: src/torch_onnx/_patch.py [hunk 8]
```diff
@@ -47,42 +47,57 @@ def _from_dynamic_axes_to_dynamic_shapes(
 
     """
     # https://github.com/pytorch/pytorch/pull/128371
+    
+    # 1. The function does not need to provide dynamic_shapes to torch.export.export
```
Comment (justinchuby) [inline src/torch_onnx/_patch.py]:
```suggestion
    # 1. The function does not need to provide dynamic_shapes to torch.export.export
```

File: src/torch_onnx/_patch.py [hunk 9]
```diff
@@ -47,42 +47,56 @@ def _from_dynamic_axes_to_dynamic_shapes(
...
+    if input_names is None:
+        input_names = []
 
-    dynamic_shapes = {}
+    sig = _signature(model)
+    if len(input_names) > len(sig.parameters):
+        raise ValueError(
+            f"Number of input names ({len(input_names)}) should not be greater than the number of model inputs ({len(sig.parameters)})"
+        )
+
```
Comment (justinchuby) [inline src/torch_onnx/_patch.py]:
```suggestion

```
